### ISSUE TITLE
Request: Backface depth

### ISSUE THREAD

**Issue by robobo1221** [sp614x/optifine#1346]
Hey sp614x,

Could we have a depth buffer for backfaces?
This would be handy for stuff like proper water fog. Since currently waterfog goes basically into infinity with depthtex1.

You can see that result here:
![image](https://user-images.githubusercontent.com/17679544/38469387-215327f4-3b54-11e8-8f1e-395482120bf3.png)
As you can see, the waterdepth basically goes to infinity instead to the backface of the water.

What the backface depth would do is act like depthtex0. But instead of including only the front face of every block, only include the backface of every transparent block.

You could make it so this depth buffer can only be used whenever backFace.translucent is set to true in shaders.properties

You could call this texture "depthtex3" or something like that. 

Thanks in regards,

**Comment by sp614x** [comment]
Water blocks have no backface where it touches solid blocks.
Not sure if "solid + water backface" depth buffer is easy to add.

**Comment by sp614x** [comment]
Can't this be done in the shaders? 
They have access to the terrain depth (depthtex2).

**Comment by robobo1221** [comment]
Yeah this can also be done in the shader. However, I do not know if that's faster than doing it in optifine. Alot of time, we use the depthbuffers in a loop. And having two texture lookups wouldn't be very practical compared to just one. It would also be two times slower.

**Comment by robobo1221** [comment]
Using two lookups is slower in general actually

**Comment by soulshot96** [comment]
gib

**Comment by dotModded** [comment]
The big thing here is not using more color attachments then we need, could we separate the depth when backfaces are enabled and write it to a buffer, yes. However doing so is incomvienient and adding read write buffers slows down the shader more than you could imagine. Not to mention being terribly inconvenient to set up. Being able to know how thick something is is very useful for absorbtion, some light scattering, and even refractions.

**Comment by sp614x** [comment]
This would need a second translucent render pass, another copy of the depth buffer and another sampler.
All 16 composite samplers are already used: https://github.com/sp614x/optifine/blob/master/OptiFineDoc/doc/shaders.txt#L215

**Comment by robobo1221** [comment]
So after some thinking I think I came up with an idea:

Make another boolean in the shader.properties called something like: useBackfaceDepthtex
What you could do instead of creating a whole new sampler uniform is to actually change depthtex1
to the backface depth. Basically how depthtex1 already is but instead of not rendering transparents at all, only render the backface of them, Also ofcourse include the solid depth.

This way you don't have to add any buffers. Plus we need to have the backface anyway for everything that uses depthtex1. 

I think this way we keep compatibility with older packs aswell.

**Comment by jcm2606** [comment]
That unfortunately wouldn't work, as shading opaque blocks requires depthtex1 to represent the depth of the opaque blocks themselves, not the possible backfaces of transparent blocks. Changing depthtex2 to represent opaque blocks would work; though you wouldn't be able to calculate a hand mask from the difference in depth between depthtex1 and depthtex2, you can solve this via a dedicated hand material mask. Aside from repurposing depthtex2, the only other option is to add another depth texture.

**Comment by robobo1221** [comment]
Hmm yeah. The best solution is to use depthtex2
